src/dali/plugin/pair_converter/ccxt.py

```diff
@@ -0,0 +1,337 @@
+# Copyright 2022 macanudo527
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from datetime import datetime, timedelta
+from time import sleep, time
+from typing import Any, Dict, List, NamedTuple, Optional, Union
+
+from ccxt import (
+    DDoSProtection,
+    Exchange,
+    ExchangeError,
+    ExchangeNotAvailable,
+    NetworkError,
+    RequestTimeout,
+    binance,
+    kraken,
+    liquid,
+)
+from rp2.logger import create_logger
+from rp2.rp2_decimal import RP2Decimal
+
+from dali.abstract_pair_converter_plugin import AbstractPairConverterPlugin
+from dali.historical_bar import HistoricalBar
+
+# Native format keywords
+_ID: str = "id"
+_BASE: str = "base"
+_QUOTE: str = "quote"
+
+# Time in ms
+_MINUTE: str = "1m"
+_FIVE_MINUTE: str = "5m"
+_FIFTEEN_MINUTE: str = "15m"
+_ONE_HOUR: str = "1h"
+_FOUR_HOUR: str = "4h"
+_ONE_DAY: str = "1d"
+_TIME_GRANULARITY: List[str] = [_MINUTE, _FIVE_MINUTE, _FIFTEEN_MINUTE, _ONE_HOUR, _FOUR_HOUR, _ONE_DAY]
+_TIME_GRANULARITY_IN_SECONDS: List[int] = [60, 300, 900, 3600, 14400, 86400]
+
+# Currently supported exchanges
+_BINANCE: str = "Binance.com"
+_KRAKEN: str = "Kraken"
+_LIQUID: str = "Liquid"
+_FIAT_EXCHANGE: str = "Exchangerate.host"
+_EXCHANGE_DICT: Dict[str, Any] = {_BINANCE: binance, _KRAKEN: kraken, _LIQUID: liquid}

```

Comment:
it looks like coinbase + coinbase pro are supported below, right?

Reply:
Basically, any CCXT supported exchange will work. You just have to add it to these key words. `_EXCHANGE_DICT` contains a `str` key which is the name of the exchange as labeled by its input plugin, and then the value is the name of the class in CCXT that handles that exchange.

So, you can add Coinbase / Coinbase Pro, or whatever they are calling themselves now to that list.

Reply:
Should we just add all supported exchanges to the list, or is there a concern in doing that?

Reply:
Well, CCXT supports over 100 exchanges, so it would take a little extra time, but we could do that. The main reason why I didn't is because lines 53~56 basically convert the `str` label from the plugins to the class used in CCXT, and since the plugins haven't been made yet, it might cause issues in the future. But, maybe we can document the 'official' name of each exchange somewhere? That would all take a lot of time, though.

Reply:
Ok, thanks for the explanation, let's go with the conservative approach you started with.